execution/execution.c

```diff
@@ -0,0 +1,219 @@
+/* ************************************************************************** */
+/*                                                                            */
+/*                                                        :::      ::::::::   */
+/*   execution.c                                        :+:      :+:    :+:   */
+/*                                                    +:+ +:+         +:+     */
+/*   By: asyed <asyed@student.42.fr>                +#+  +:+       +#+        */
+/*                                                +#+#+#+#+#+   +#+           */
+/*   Created: 2018/03/21 21:16:12 by asyed             #+#    #+#             */
+/*   Updated: 2018/04/04 17:34:58 by asyed            ###   ########.fr       */
+/*                                                                            */
+/* ************************************************************************** */
+
+#include "exec.h"
+#include <stdio.h>
+#include <stdlib.h>
+#include <strings.h>
+#include <errno.h>
+
+struct s_ophandlers	op_handlers[] = {
+	{&op_pipe_check, &op_pipe_exec},
+	{&op_or_check, &op_or_exec},
+	{&op_and_check, &op_and_exec},
+	{NULL, NULL},
+};
+
+int		run_operation(t_ast *curr, uint8_t wait)
+{
+	pid_t	pid;
+	int 	res;
+
+	if (!curr || *(curr->type) == operator)
+		return (EXIT_FAILURE);
+	if ((res = builtin_handler(curr)) != -1)
+		return (res);
+	if ((pid = fork()) == -1)
+		return (EXIT_FAILURE);
+	if (pid == 0)
+	{
+		if (handle_redirection(curr))
+			exit(EXIT_FAILURE);
+		printf("[In: %d, Out: %d, Err: %d]\n", *(curr->p_info->stdin), *(curr->p_info->stdout), *(curr->p_info->stderr));
+		dup2(*(curr->p_info->stdin), STDIN_FILENO);
+		dup2(*(curr->p_info->stdout), STDOUT_FILENO);
+		dup2(*(curr->p_info->stderr), STDERR_FILENO);
+		execvP(*(curr->token), getenv("PATH"), curr->token);
+		exit(EXIT_FAILURE);
+	}
+	if (wait)
+	{
+		waitpid(pid, &res, 0);
+		if (!res)
+			return (EXIT_SUCCESS);
+		else
+			return (EXIT_FAILURE);
+	}
+	return (EXIT_SUCCESS);
+}
+
+static inline __attribute__((always_inline)) void	*init_process(void)
+{
+	t_process	*new;
+
+	if (!(new = ft_memalloc(sizeof(t_process))))
+		return (NULL);
+	else if (!(new->stdin = ft_memalloc(sizeof(int))))
+	{
+		free(new);
+	}
+	else if (!(new->stderr = ft_memalloc(sizeof(int))))
+	{
+		free(new);
+		free(new->stdin);
+	}
+	else if (!(new->stdout = ft_memalloc(sizeof(int) * 2)))
+	{
+		free(new);
+		free(new->stdin);
+		free(new->stderr);
+	}
+	else if (!(new->comm = ft_memalloc(sizeof(int) * 2)))
+	{
+		free(new);
+		free(new->stdin);
+		free(new->stderr);
+		free(new->stdout);
+	}
+	else
+		return (new);
+	return (NULL);
+}
+
+void	build_leafs(t_ast *curr)
```

Comment:
Even if i should never occur, make sure curr is not null for consistency, cuz u did it for some functions